### PR TITLE
simulator function accounting for susceptible depletion

### PR DESCRIPTION
**Pull request by ffinger**
Hi @sbfnk,
I created a version of your simulator that can take into account the depletion of susceptibles in a finite population. It only works with poisson and negative binomial distributions so far, they are truncated to avoid overshooting the remaining susceptibles. As it needs to keep track of the remaining susceptibles each time the offspring is drawn I simulate the draws one by one. Let me know if you have a more elegant idea.
Cheers.

### CODE REVIEW

File: R/simulate_susceptibles.R [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    }
+
+    if (offspring == "pois") {
+        ## using a right truncated poisson distribution
+        ## to avoid more cases than susceptibles
+        offspring_fun <- function(n, susc) {
+            extraDistr::rtpois(
+                n,
+                lambda = mn_offspring * susc / pop,
+                a = 0,
```
Comment (sbfnk) [inline R/simulate_susceptibles.R]: This seems to exclude 0. Should it be `a=-Inf` (or just don't set `a`)?

File: R/simulate_susceptibles.R [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+            ## see ?rnbinom for parameter definition
+            new_mn <- mn_offspring * susc / pop ##apply susceptibility
+            size <- new_mn / (disp_offspring - 1)
+
+            ## using a right truncated nbinom distribution
+            ## to avoid more cases than susceptibles
+            truncdist::rtrunc(
+                n,
+                spec = "nbinom",
+                a = 0,
```
Comment (sbfnk) [inline R/simulate_susceptibles.R]: Same as above - remove this line?

File: R/simulate_susceptibles.R [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+            tdf <- rbind(tdf, new_df)
+        }
+
+        ## adjust susceptibles
+        susc <- susc - n_offspring
+    }
+
+    ## remove cases with time > tf that could
+    ## have been generated in the last generation
+    tdf2 <- tdf[tdf$time <= tf, ]
```
Comment (sbfnk) [inline R/simulate_susceptibles.R]: should be `tdf <-`? Or change the occurences below to `tdf2`

File: R/simulate_susceptibles.R [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+##'   the R distribution function. Currently only "pois" & "nbinom" are
+##'   supported. Internally truncated distributions are used to avoid infecting
+##'   more people than susceptibles available.
+##' @param mn_offspring the average number of secondary cases for each case
+##' @param disp_offspring the dispersion coefficient (var/mean) of the number of
+##'      secondary cases. Ignored if offspring == "pois". Must be > 1.
+##' @param serial the serial interval. A function that takes one parameter
+##'     (`n`), the number of serial intervals to randomly sample.
+##'     Value must be >= 0.
+##' @param t0 start time (if serial interval is given); either a single value (0
```
Comment (sbfnk) [inline R/simulate_susceptibles.R]: For parameters that are exactly the same (with the same description) as in `chain_sim` you could remove the documentation here and instead insert a line `##' @inheritParams chain_sim` below

File: R/simulate_susceptibles.R [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        ## adjust susceptibles
+        susc <- susc - n_offspring
+    }
+
+    ## remove cases with time > tf that could
+    ## have been generated in the last generation
+    tdf2 <- tdf[tdf$time <= tf, ]
+
+    ## sort output and remove columns not needed
+    tdf <- dplyr::arrange(tdf, time, id)
```
Comment (sbfnk) [inline R/simulate_susceptibles.R]: This adds the `dplyr` dependency just to sort a data frame. Is `dplyr::arrange` faster than `base::order`? If not we could use 
```r
tdf <- tdf[order(tdf$time, tdf$id), ]
```

File: R/utils.r [hunk 6]
```diff
@@ -57,3 +57,18 @@ find_function_name <- function(fun) {
...
+##' in terms of mean and coefficient of variation
+##' @param n number of samples to draw
+##' @param mn mean of distribution
+##' @param disp dispersion coefficient (var/mean)
+##' @return vector containing the random numbers
+##' @author Flavio Finger
+##' @export
+##' @examples
+##' rnb_mean_disp(n = 5, mn = 4, disp = 2)
+rnb_mean_disp <- function(n, mn, disp) {
```
Comment (sbfnk) [inline R/utils.r]: Minor comment - shall we call it `rnbinom_mean_disp` to be more similar to `rnbinom`?

**Comment by ffinger** [comment]
Thanks for the very useful reviews and for pointing out the few flaws, I changed those and also implemented some tests.

**Comment by ffinger** [comment]
I squashed the previous series of commits into just 2 to get a cleaner history.

**Comment by sbfnk** [comment]
Two final requests:
1. Do we need the `extraDist` dependency? Or can we sample from both distributions using `truncdist::rtrunc`?
2. Some of the tests for errors fail sometimes because the first generation generates 0 offspring and the `while` loop is never entered. I can't think of any obvious solution for this, so we may have to remove these tests for now (or else we could run into all sorts of problems with CRAN checks failing etc.).

**Comment by ffinger** [comment]
> Two final requests:
> 
>     1. Do we need the `extraDist` dependency? Or can we sample from both distributions using `truncdist::rtrunc`?

I think we could rely on truncdist::rtrunc only, yes. I can change this.

> 2. Some of the tests for errors fail sometimes because the first generation generates 0 offspring and the `while` loop is never entered. I can't think of any obvious solution for this, so we may have to remove these tests for now (or else we could run into all sorts of problems with CRAN checks failing etc.).

Ok, I see, the one for the serial interval is inside an ìf` statement that doesn't execute if there is 0 offspring. Good spot.

**Comment by sbfnk** [comment]
Thanks @ffinger for a great contribution!

**Pull request merged**